File: python/command_injection_vulnerable.py
"""
Vulnerable Python code with command injection flaws.
For security scanner testing only.
"""
import os
import subprocess


def ping_host_vulnerable(hostname):
    """Command injection via os.system."""
    os.system(f"ping -c 4 {hostname}")


def execute_command_vulnerable(cmd):
    """Command injection via subprocess with shell=True."""
    subprocess.call(cmd, shell=True)


def run_script_vulnerable(script_name):
    """Command injection via subprocess.run with shell=True."""
    subprocess.run(f"python {script_name}", shell=True)


def check_file_vulnerable(filename):
    """Command injection via os.popen."""
    result = os.popen(f"ls -la {filename}").read()
    return result


def process_file_vulnerable(filepath):
    """Command injection via subprocess.Popen."""
    subprocess.Popen(f"cat {filepath}", shell=True, stdout=subprocess.PIPE)


def eval_vulnerable(user_code):
    """Arbitrary code execution via eval."""
    result = eval(user_code)
    return result


def exec_vulnerable(user_code):
    """Arbitrary code execution via exec."""
    exec(user_code)


def ping_host_safe(hostname):
    """Safe version using argument list."""
    subprocess.run(['ping', '-c', '4', hostname], check=True)


def run_command_safe(filename):
    """Safe version with proper argument passing."""
    subprocess.run(['ls', '-la', filename], check=True)

File: javascript/command_injection_vulnerable.js
/**
 * Vulnerable JavaScript code with command injection flaws.
 * For security scanner testing only.
 */

const { exec, execSync, spawn } = require('child_process');

function pingHostVulnerable(hostname) {
    // Command injection via exec
    exec(`ping -c 4 ${hostname}`, (error, stdout, stderr) => {
        console.log(stdout);
    });
}

function runCommandVulnerable(command) {
    // Command injection via execSync
    const output = execSync(command);
    console.log(output.toString());
}

function processFileVulnerable(filename) {
    // Command injection via exec with user input
    exec('cat ' + filename, (error, stdout, stderr) => {
        console.log(stdout);
    });
}

function listFilesVulnerable(directory) {
    // Command injection via template literal
    const output = execSync(`ls -la ${directory}`);
    return output.toString();
}

function evalVulnerable(userCode) {
    // Arbitrary code execution via eval
    const result = eval(userCode);
    return result;
}

function functionConstructorVulnerable(userCode) {
    // Arbitrary code execution via Function constructor
    const fn = new Function(userCode);
    return fn();
}

function pingHostSafe(hostname) {
    // Safe version using spawn with argument array
    const child = spawn('ping', ['-c', '4', hostname]);
    child.stdout.on('data', (data) => {
        console.log(data.toString());
    });
}

function processFileSafe(filename) {
    // Safe version using spawn
    const child = spawn('cat', [filename]);
    child.stdout.on('data', (data) => {
        console.log(data.toString());
    });
}
